src/common/eigf.h

```diff
@@ -0,0 +1,302 @@
+ * the other (dividing by average^2 smoothed too much dark details surrounded
+ * by bright pixels).
+ *
+ * This modification makes the filter exposure-independent.
+ * However, due to the fact that the average advantages the bright pixels
+ * compared to dark pixels if we consider that human eye sees in log,
+ * we get strong bright halos.
+ * These are due to the spatial averaging of "a" and "b" that is performed at
+ * the end of the filter, especially due to the spatial averaging of "b".
+ * However, removing completely this averaging gives results which are not
+ * smoothed enough.
+ * Hence, we use a weighted averaging of "a" and "b" to overcome this problem.
+ * We weight each "a" and "b" by 1 / pixel_value^2, which gives much less
+ * halos problems, and gives a smooth result.
+ * Weighting by 1 / pixel_value^2 is empirical.
+ * It was found to work well, while weighting by 1/pixel_value gives sometimes
+ * worse results.
+ * The idea of using a weighted averaging here comes from:
+ * C. N. Ochotorena and Y. Yamashita, "Anisotropic Guided Filtering," in IEEE Transactions on Image Processing, vol. 29, pp. 1397-1412, 2020, doi: 10.1109/TIP.2019.2941326.
+ * although we don't use the same weighting system (their weight is based
+ * on variance, ours is based on the pixel value).
+ * We tried a mixed weight using their weight in combination to ours, but
+ * it did not improve the result.
+ *
+ * The implementation EIGF uses downscaling to speed-up the filtering,
+ * just like what is done in fast_guided_filter.h
+**/
+
+static inline void exposure_independent_guided_filter(const float *const restrict guide, // I
+                                    const float *const restrict mask, //p
+                                    float *const restrict ab,
+                                    const size_t width, const size_t height,
+                                    const float sigma, const float feathering)
+{
+  // We also use gaussian blurs instead of the square blurs of the guided filter
+  const size_t Ndim = width * height;
+  float *const restrict blurred_guide = dt_alloc_sse_ps(Ndim);
+  // guide_x_guide = (guide - blurred_guide)^2
+  float *const restrict guide_x_guide = dt_alloc_sse_ps(Ndim);
+  // guide_variance = blur(guide_x_guide)
+  float *const restrict guide_variance = dt_alloc_sse_ps(Ndim);
+  float *const restrict guide_x_mask = dt_alloc_sse_ps(Ndim);
+  // guide_mask_covariance = blur(guide_x_mask)
+  float *const restrict guide_mask_covariance = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_mask = dt_alloc_sse_ps(Ndim);
+  float *const restrict a = dt_alloc_sse_ps(Ndim);
+  float *const restrict b = dt_alloc_sse_ps(Ndim);
+  // weight to compute the weighted blur of a and b
+  float *const restrict weights = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_a = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_b = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_weights = dt_alloc_sse_ps(Ndim);
```

Comment:
Can't you reduce the number of memory buffers ? In fast_guided_filter.h, I use :
```
  const size_t Ndim = width * height;
  const size_t Ndimch = width * height * 4;

  float *const restrict temp = dt_alloc_sse_ps(Ndimch); // array of structs { { mean_I, mean_p, corr_I, corr_Ip } }
  float *const restrict guide_x_mask = dt_alloc_sse_ps(Ndim);
  float *const restrict guide_x_guide = dt_alloc_sse_ps(Ndim);
```
and output an `{a, b}` vector that gets blurred in a single loop × 2 channels. Everything else is computed in-loop, with the blur merged in there.

Reply:
Actually, to use gaussian blur I need either a 1 channel buffer, or a 4 channels buffer. That's why I have a and b in separate buffers